include/rmm/mr/device/cuda_async_memory_resource.hpp

```diff
@@ -0,0 +1,149 @@
+    cudaMemPool_t pool;
+    RMM_ASSERT_CUDA_SUCCESS(cudaDeviceGetDefaultMemPool(&pool, device));
+    RMM_ASSERT_CUDA_SUCCESS(cudaMemPoolTrimTo(pool, 0));
+  }
+
+  cuda_async_memory_resource(cuda_async_memory_resource const&) = default;
+  cuda_async_memory_resource(cuda_async_memory_resource&&)      = default;
+  cuda_async_memory_resource& operator=(cuda_async_memory_resource const&) = default;
+  cuda_async_memory_resource& operator=(cuda_async_memory_resource&&) = default;
+
+  /**
+   * @brief Query whether the resource supports use of non-null CUDA streams for
+   * allocation/deallocation. `cuda_memory_resource` does not support streams.
+   *
+   * @returns bool true
+   */
+  bool supports_streams() const noexcept override { return true; }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return true
+   */
+  bool supports_get_mem_info() const noexcept override { return true; }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using cudaMalloc.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, rmm::cuda_stream_view stream) override
+  {
+    void* p{nullptr};
+    if (bytes > 0) { RMM_CUDA_TRY(cudaMallocAsync(&p, bytes, stream.value()), rmm::bad_alloc); }
+    return p;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   */
+  void do_deallocate(void* p, std::size_t, rmm::cuda_stream_view stream) override
+  {
+    if (p != nullptr) { RMM_ASSERT_CUDA_SUCCESS(cudaFreeAsync(p, stream.value())); }
+  }
+
+  /**
+   * @brief Compare this resource to another.
+   *
+   * @throws Nothing.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   */
+  bool do_is_equal(device_memory_resource const& other) const noexcept override
+  {
+    return dynamic_cast<cuda_async_memory_resource const*>(&other) != nullptr;
+  }
+
+  /**
+   * @brief Get free and available memory for memory resource
+   *
+   * @throws `rmm::cuda_error` if unable to retrieve memory info.
+   *
+   * @return std::pair contaiing free_size and total_size of memory
+   */
+  std::pair<size_t, size_t> do_get_mem_info(rmm::cuda_stream_view) const override
+  {
+    std::size_t free_size;
+    std::size_t total_size;
+    RMM_CUDA_TRY(cudaMemGetInfo(&free_size, &total_size));
+    return std::make_pair(free_size, total_size);
+  }
+};
+
+#else  // CUDART_VERSION < 11020 (11.2)
+
+#include <rmm/mr/device/cuda_memory_resource.hpp>
+
+using cuda_async_memory_resource = rmm::mr::cuda_memory_resource;
```

Comment:
Are we sure we want to do this? Things like `supports_streams` would return differently which means someone would potentially need to special case all usage of `cuda_async_memory_resource`. Should we just have it throw at runtime if someone tries to create a `cuda_async_memory_resource`?

Reply:
We can't just throw at runtime, we also have to disable it at compile time. I can just remove this alias, but then I have to modify all of my tests and benchmarks to also check at compile time.

`supports_streams` is going away. Nobody uses it and we are redesigning the base interface.

Reply:
> I can just remove this alias, but then I have to modify all of my tests and benchmarks to also check at compile time.

Everyone who builds against RMM who wants to support multiple CUDA versions in a single binary and potentially use `cuda_async_memory_resource` would also have to do the same thing then which is arguably a bit burdensome?

My thought is throwing at runtime would presumably allow someone to choose the best allocator that's available for them at runtime.

Reply:
I don't fully understand. Are you suggesting in the non-CUDA11.2-compilation path to define a class `cuda_async_memory_resource` that just throws in the constructor? 

This does have the problem that tests and benchmarks also have to check for compatibility before trying to construct it, but OK.

Reply:
> Are you suggesting in the non-CUDA11.2-compilation path to define a class cuda_async_memory_resource that just throws in the constructor?

Yes, but in my ideal world it's a runtime path as opposed to a compilation path so that a user can upgrade their CUDA toolkit runtime version without needing to rebuild the entire dependency tree below RMM.

Then at runtime they could do something like:
```
try:
    mr = cuda_async_memory_resource()
except UnsupportedCUDARuntimeVersion or UnsupportedCUDADriverVersion:
    mr = some_other_memory_resource()
    
...
```

Then we end up with the following scenarios:
- Built with CUDA 11.2+
  - Running with CUDA 11.2+ runtime and driver: `cuda_async_memory_resource` is used
  - Running with < CUDA11.2 runtime OR driver: `some_other_memory_resource` is used
- Built with CUDA 11.0
  - Running with CUDA 11.2+ runtime and driver: `cuda_async_memory_resource` is used
  - Running with < CUDA 11.2 runtime or driver: `some_other_memory_resource` is used

Reply:
With the < CUDA 11.2 compilation path option it doesn't give the option of upgrading the toolkit to a new version and allowing new features to be tapped into without rebuilding / pulling new builds. We could be perfectly fine making this trade-off to reduce the development / maintenance burden, but for the best user experience we'd enable forwards compatibility without needing to rebuild.

Reply:
> Built with CUDA 11.0 runtime
>   Running with CUDA 11.2+ runtime and driver: cuda_async_memory_resource is used

This one is not possible. It's not possible for `cuda_async_memory_resource` to function if it was not built with CUDA 11.2+.

Reply:
OK, so I changed it to be a runtime exception in all cases where cudaMallocAsync was not available when compiled. Also added a test for this.  That said, this is the behavior you will be able to achieve (3/4 of what you requested):


    Built with CUDA 11.2+ runtime
        Running with CUDA 11.2+ runtime and driver: cuda_async_memory_resource is used
        Running with < CUDA11.2 runtime OR driver: some_other_memory_resource is used
    Built with CUDA 11.0 runtime
        some_other_memory_resource is used

Reply:
> This one is not possible. It's not possible for `cuda_async_memory_resource` to function if it was not built with CUDA 11.2+ because the APIs aren't there for it to call.

Isn't it possible with the `dlopen` / `dlsym` approach since symbols are resolved at runtime as opposed to compile time? The development cost is that we'd have to likely create our own signature for `cudaMallocAsync` and the runtime cost is we'd dynamically load the library and have to version check somewhere.

Reply:
Also, I'm perfectly fine if the answer is this is something we don't want to support, but I want to make sure we have proper discussion before we arrive at that conclusion.

Reply:
Maybe. My goal was to avoid that. With this simpler approach, by building our CUDA 11.x artifacts going forward using CUDA 11.2, the majority of our users get the behavior you desire. Only those who build from source with an old CUDA version don't get it.

Reply:
Fair enough. Would welcome other thoughts here before we merge, but that seems like a completely reasonable tradeoff to me.

cc @rongou @germasch @jrhemstad for thoughts